### ISSUE TITLE
Some details about ur matching strategy

### ISSUE THREAD

**Issue by takecareofbigboss** [sfzhang15/SFD#12]
@sfzhang15 hi, i have implemented the code of data augmentation & maxout in your training process, and just use these(change some codes in SSD), our model can achieve 93.6/ 92.3/ 84.0. But when i implemented anchor matching in function matchbbox(), the performance drops. Can you guide me how to solve this problem?  
I just add step 2 after Bipartite matching(step 1 is the source code of ssd, i donot changed it). What surprised me is that i found after step 1, above 10 anchors of each gt faces is chosed (not below 5 as you said). Maybe i should change some code in step 1?

**Comment by takecareofbigboss** [comment]
any other tricks in your paper? 
i suppose that the revised code i mentioned has no errors, maybe there remains some hyperparameters to be changed?

**Comment by takecareofbigboss** [comment]
@sfzhang15  Hope i can get ur reply ASAP.

**Comment by sfzhang15** [comment]
@takecareofbigboss Hi, 10 anchors of each gt faces is for IoU threshold = 0.35 and below 5 is for IoU threshold = 0.5, this is the why we change the IoU threshold from 0.5 to 0.35.
After step 1, for each gt whose number of matched anchors is less than 6, we use the step 2.

**Comment by takecareofbigboss** [comment]
1. Will you use IoU threshold = 0.5 to pre-train your model or just directly use IoU threshold = 0.35?
2. For IoU threshold = 0.35, if the number of anchors of each gt faces is above 10, we choose them all or just choose top10?
3. i dont undestand how do you deal with 6 and 10. (You mean you set N=6?)

**Comment by takecareofbigboss** [comment]
i am confused about what you said, and i suppose it is conflicting.
"10 anchors of each gt faces is for IoU threshold = 0.35"
&"After step 1, for each gt whose number of matched anchors is less than 6, we use the step 2."

**Comment by sfzhang15** [comment]
1. We just directly use IoU threshold = 0.35
2. choose them all
3. After step 1, for these GT whose matched anchors are less than 6, we use step 2 to choose its top6 anchor to match

**Comment by jianyin2016** [comment]
hello,zhang!
1,I am puzzled by some details of the matching strategy you have mentioned. as you described in the paper that " We deal with each of these faces as follow: firstly picking out anchors whose jaccard overlap with this face are higher than 0.1, then sorting them to select top-N as matched anchors of this face. We set N as the average number from stage one".
I wonder the top-N anchors with IOU>0.1 should be the same as the top-N anchors with positive IOU，so why the param 0.1 was necessary？

2,Besides,as for the implementation of stage 2 in MatchBBox() ,are you just change the elements which correspond to the top-N overlapping anchors  of  vector match_indices from -1 to corresponding GT index j ,so you implement a anchor number count function and a rank IOU function to support your implementation,maybe i can do it with tons of for loops,but I wonder is there any more efficient functions to  contribute?

A newcomer in detection and a freshman in C++,feel free to correct me if I am wrong or not clearly clarifying my question ,many many thanks!

**Comment by sfzhang15** [comment]
@jianyin2016 Hi,

1. Since tiny faces always has some matched anchors with less then 0.1 IoU and the param 0.1 is used to filter out those too biased anchor.  The N of top-N is the average number of matched anchors from stage one, we set it as 6.

2. You can use 'multimap' to store overlaps as <int j, <float overlap, int i>>.

**Comment by jianyin2016** [comment]
@sfzhang15 Thanks for your reply!
 I must ensure I haven't wrongly understanding your design，I think the top-6 tiny face anchors must have their IoUs larger than 0.1，as at the beginning of match ,we have save all the positive using the multimap you said:
if (overlap > 1e-6) {
        (*match_overlaps)[i] = std::max((*match_overlaps)[i], overlap);
        overlaps[i][j] = overlap;
      }
the restriction of IoU>0.1 doesn't filter out the top-N anchors,but only reduces the amount of alternative anchors you are going to sort? even you choose the IoU threshold 0.0,the top-N anchors unchanges,just the time consumption  reduces，am I right？

**Comment by sfzhang15** [comment]
@jianyin2016 
1. The top-6 tiny face anchors must have their IoUs larger than 0.1
2. You should use **'multimap <int, <float, int>>'** to save all the positive whose iou is large than 0.1
3. Lots of tiny faces has less than 6 matched anchors and many of these 6 anchors has IoU < 0.1, so the restriction of IoU>0.1 will promise that those anchors whose IoU < 0.1 will not be selected.

**Comment by jianyin2016** [comment]
get it！many many thanks~

**Comment by abdelrahman-gaber** [comment]
@takecareofbigboss Can you please tell me if you did the modification of augmentation in the train.prototxt file or we need to change something inside the C++ code of Caffe, if so, which file should we change? 
and can you please share the code you did about the modifications in augmentation and matching strategy? It would be really appreciated.

**Comment by jnulzl** [comment]
@takecareofbigboss @jianyin2016 @abdelrahman-gaber ,I try to implement the matching strategy(only cpu version). But I have not debug.Maybe you can  help me testing.Below is my code, just insert to the last of function MatchBBox in bbox_utils.cpp.
`
//********************Stage Two*****************
switch (match_type) {
    case MultiBoxLossParameter_MatchType_BIPARTITE:
      // Already done.
      break;
    case MultiBoxLossParameter_MatchType_PER_PREDICTION:
      {
		vector<int> pred_for_gt_idx_count;
		pred_for_gt_idx_count.resize(num_gt,0);
		int all_matched_pred = 0;
		for(int i=0;i < num_pred;i++){
			int temp_gt_idx = (*match_indices)[i];
			if(temp_gt_idx >=0) {
				pred_for_gt_idx_count[temp_gt_idx]++;
				all_matched_pred++;
			}
		}
		int N = all_matched_pred / num_gt;
		for(int k=0;k < num_gt;k++){
			if(pred_for_gt_idx_count[k] >= N) {
				continue;
			}
			map<float,int,std::greater<float>> gt_k_and_pred;
			for(map<float,map<int,float> >::iterator it = overlaps.begin();it != overlaps.end();++it){
				int i = it->first;
				if((*match_indices)[i] != -1  || it->second.find(k) == it->second.end()){
					continue;
				}
				float overlap = it->second[k];
				if(overlap >= 0.1) {
					gt_k_and_pred[overlap] = i;
				}
			}
			for(map<float,int,std::greater<float>>::iterator it = gt_k_and_pred.begin(),int j = 0;it != gt_k_and_pred.end() && j < N;++it,++j){
		(*match_indices)[it->second]  = k;
		(*match_overlaps)[it->second]  = it->first;
	}
		}
	}
      break;
    default:
      LOG(FATAL) << "Unknown matching type.";
      break;
  }
`

**Comment by jnulzl** [comment]
@takecareofbigboss How to implement data augmentation？Share your code?

**Comment by fatporkchop** [comment]
@jnulzl Have you debugged your code? Can please share your code?
